lib/Zonemaster/Backend/DB.pm

```diff
@@ -130,6 +133,84 @@ sub process_unfinished_tests {
     }
 }
 
+sub _normalize_parameter_hash {
+    my ( $self, $params ) = @_;
+
+    my $profile = Zonemaster::Engine::Profile->effective;
+
+    my %normalized = ();
+    $normalized{domain}   = lc $$params{domain} // "";
+    $normalized{ipv4}     = $$params{ipv4}      // $profile->get( 'net.ipv4' );
+    $normalized{ipv6}     = $$params{ipv6}      // $profile->get( 'net.ipv6' );
+    $normalized{profile}  = $$params{profile}   // "default";
+    $normalized{priority} = $$params{priority}  // 10;
+    $normalized{queue}    = $$params{queue}     // 0;
+
+    my $array_ds_info = $$params{ds_info} // [];
+    my @array_ds_info_sort = sort {
+        (
+            defined $a->{algorithm} and
+            defined $b->{algorithm} and
+            $a->{algorithm} cmp $b->{algorithm}
+        ) or
+        (
+            defined $a->{digest} and
+            defined $b->{digest} and
+            $a->{digest} cmp $b->{digest}
+        ) or
+        (
+            defined $a->{digtype} and
+            defined $b->{digtype} and
+            $a->{digtype} <=> $b->{digtype}
+        ) or
+        (
+            defined $a->{keytag} and
+            defined $b->{keytag} and
+            $a->{keytag} <=> $b->{keytag}
+        )
+    } @$array_ds_info;
+
+    $normalized{ds_info} = \@array_ds_info_sort;
+
+    my $array_nameservers = $$params{nameservers} // [];
+    my @array_nameservers_sort = sort {
+        (
+            defined $a->{ip} and
+            defined $b->{ip} and
+            $a->{ip} cmp $b->{ip}
+        ) or
+        (
+            defined $a->{ns} and
+            defined $b->{ns} and
+            $a->{ns} cmp $b->{ns}
+        )
+    } @$array_nameservers;
```

Comment:
A nameserver object can be both `{"nameservers":[{"ns":"ns.nic.se","ip":""}` and `{"nameservers":[{"ns":"ns.nic.se"}`. Will your code normalize that? -- Maybe that is too ambitious.

Reply:
Empty `ip` and omitted `ip` mean the same thing, right? In that case we should be normalized.

Reply:
Done with 10a5dca.